design/build-clusters.md

```diff
@@ -0,0 +1,163 @@
+
+
+### Queue setup
+TBD
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```
+
+
+### New table for build cluster details
+
+Table: `buildCluster`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *false or true* |
+| `defaultCluster` | boolean | no | no | no | *true - cluster maintained by screwdriver team, false - cluster maintained by  external team * |
+| `maintainer` | text(100) | yes | no | no | cluster admin email for communications |
+
+Unique constraint: `name` 
+
+#### Sample record
+
+| id | name | scmContext | scmOrganizations | isActive | defaultCluster | maintainer 
+| --- | --- | --- | --- | --- | --- | --- | 
+| 1 | gq1 | github:github.com | null | true | true | sd@foo.com |
+| 2 | bf1 | github:github.com | null | false | true | sd@foo.com |
+| 3 | iOS | github:github.com | [iOS_org1, iOS_org2] | true | false | ios@foo.com |
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url | Description
+| --- | --- | ---
+| `POST` | ` /buildClusters ` | ` { "name":"iOS", "scmContext":"github:github.com", "scmOrganizations": "[iOS_org1, iOS_org2]", "isActive":true, "defaultCluster": true } `
+| `GET` | `	/buildClusters ` | ` get list of buildClusters info `
+| `GET` | `	/buildClusters/:name ` | ` get a particular buildCluster info `
+| `DELETE` | ` /buildClusters/:name ` | ` delete buildCluster `
+
+
+### Cluster on-board
+
+1. Build cluster admin requesting access with cluster info and user credentials
+2. SD admin to populate buildCluster table with cluster info 
+3. SD admin to create queue based on #1
+4. SD admin authorize build cluster user and queue 
+
+
+## Flow
+### SD Validator
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### API to Redis
+
+	1. No change in existing implementation.
+
+### Model
+
+	1. Query `buildClusters` table for active records with cluster name from build info 
+	2. Validates if build job can be scheduled in specified buildCluster queue by validating scmContext + scmOrganization access. 
+	3. one (or) more record exist, then assign job to the queue identified by generating a random number within given boundaries, which is the returned list size of records
```

Comment:
Also we need to capture that this works as long as build nodes in cluster is homogeneous. We should also look to add weighs to individual clusters if there is significant difference in capacity between them.

Reply:
consumer fetch count per cluster can be configured individually based on their capacity. multiple clusters can listen to one queue and pull messages based on the configured fetch count.

Reply:
as discussed, included weighted percentage, but percentage should be adjusted automatically (or) manually when a cluster is made inactive.